grafana-stack/deploy.sh

```diff
@@ -0,0 +1,191 @@
+        ;;
+      "influxdb")
+        influxdb=1
+        ;;
+      "nginx")
+        nginx=1
+        ;;
+      *)
+        echo "Unknown module: ${module}"
+        echo "usage: $0 [subdir...]"
+        exit 64
+    esac
+  done
+fi
+
+has() {
+  [[ $1 -ne 0 ]]
+}
+
+print_line() {
+  echo $(date +"%F-%H %T.%N"): "$@" >&3
+
+}
+
+run_command() {
+  what=$1
+  shift
+  print_line ${what}
+  "$@" || {
+    print_line Failed. You can have a look at \"deploy.log\"
+    exit 3
+  }
+}
+
+print_line "Running against modules: ${modules}"
+
+run_command "Reading configuration: deploy.conf" . ./deploy.conf
+print_line "Configuration used:"
+print_line "Kubeconfig:                 ${KUBECONFIG}"
+print_line "Influxdb root password:     ${INFLUXDB_ROOT_PASSWORD}"
+print_line "Grafana admin password:     ${GRAFANA_ADMIN_PASSWORD}"
+print_line "Server host:                ${SERVER_HOST}"
+
+exec 2>deploy.log
+exec 1>&2
+
+set -o xtrace
+
+apply_config() {
+  config=$1
+  kubectl --kubeconfig="${KUBECONFIG}" apply -f "${config}"
+}
+
+create_pd() {
+  # This is far from perfect. The size must match what is in configuration files
+  name=$1
+  vclaim=$2
+  volume=$3
+  size=$4
+
+  ! [[ -z "$(gcloud compute disks list --uri ${name})" ]] || \
+          gcloud compute disks create "${name}" --size "${size}" &&
+  apply_config "${vclaim}" &&
+  apply_config "${volume}"
+}
+
+get_public_ip() {
+  service=$1
+
+  kubectl --kubeconfig="${KUBECONFIG}" get svc "${service}" -o=jsonpath="{.status.loadBalancer.ingress[0].ip}"
+}
+
+wait_nginx_service() {
+  for i in $(seq 12); do
+    ip=$(get_public_ip nginx-service) &&
+    if [[ ! -z ${ip} ]]; then
+      print_line "Nginx service is up:" ${ip}
+      return
+    fi
+    sleep 5
+  done
+  false
+}
+
+prepare_grafana_secret() {
+  root_password=$1
+
+  kubectl --kubeconfig="${KUBECONFIG}" delete secret grafana || true
+  kubectl --kubeconfig="${KUBECONFIG}" create secret generic grafana "--from-literal=rootpassword=$root_password"
+}
+
+prepare_influxdb_secret() {
+  root_password=$1
+
+  kubectl --kubeconfig="${KUBECONFIG}" delete configmap influxdb || true
+  kubectl --kubeconfig="${KUBECONFIG}" create configmap influxdb --from-file=influxdb.toml
+
+  kubectl --kubeconfig="${KUBECONFIG}" delete secret influxdb || true
+  kubectl --kubeconfig="${KUBECONFIG}" create secret generic influxdb "--from-literal=rootpassword=$root_password"
+}
+
+prepare_nginx_configmap() {
+  kubectl --kubeconfig="${KUBECONFIG}" delete configmap nginx || true
+  kubectl --kubeconfig="${KUBECONFIG}" create configmap nginx --from-file=nginx.conf
+}
+
+create_grafana_datasource() {
+  server_hostname=${1:-$(get_public_ip nginx-service)} &&
+  grafana_admin_password=$2 &&
+  influxdb_root_password=$3 &&
+
+  curl -s --fail "http://admin:${grafana_admin_password}@${server_hostname}/api/datasources/name/github" ||
+  cat datasource.json |
+  env - server_hostname=${server_hostname} influxdb_root_password=${influxdb_root_password} envsubst |
```

Comment:
probably should put vars in quotes, especially things like `"${influxdb_root_password}"`